app/models/concerns/thredded/topic_common.rb

```diff
@@ -47,22 +47,30 @@ def unread(user)
           .merge(reads_class.where(reads[:id].eq(nil).or(reads[:read_at].lt(topics[:updated_at]))))
       end
 
+      def read_states_by_topics_lookup(user)
+        read_states_by_topic_id =
+          reflect_on_association(:user_read_states).klass
+            .where(user_id: user.id, postable_id: current_scope.map(&:id))
+            .group_by(&:postable_id)
+
+        def read_states_by_topic_id.get(topic, null_value = nil)
```

Comment:
@glebm Hash with default value won't work as we need an accessor that passes in the topic (to prevent the O(n) SQL calls on every row). Maybe the answer is to have a specific class to handle the lookup. I'll try that and you see what you think